File: python/component/src/new_world_parser.py
"""Parsing core for the ``new-world-product-search`` WIT interface
(``common/wit/html-parser.wit``).

One parsing core per language (docs/html-parser-plan.md): this module is
used directly by the host-side native tests in ``tests/`` and compiled
unchanged into ``new-world-parser.wasm`` through the thin component glue in
``parser_app.py`` -- there is never a second copy of the extraction logic
in this language.

The HTML tree is built with the stdlib ``html.parser`` tokenizer only
(componentize-py bundles the stdlib; constitution section 6.3 -- no
third-party parser dependency). ``parse_search_results`` returns the
harness-wide result envelope directly -- ``{"ok": {...}}`` or
``{"err": {...}}`` -- with every record as a plain dict keyed exactly like
the JSON contract (kebab-case, ``None`` for an absent WIT ``option<>``).

URL resolution is a small explicit RFC-3986-style resolver implemented
identically in all three language cores (rather than each language's URL
library) so canonical URLs -- and therefore deduplication keys -- are
byte-identical across implementations.
"""

from __future__ import annotations

from html.parser import HTMLParser

_SITE = "new-world"
_PRODUCT_ITEMTYPE = "https://schema.org/Product"
_PRODUCT_TESTID_PREFIX = "product-"
_U32_MAX = 4294967295

# The element that makes a document recognizable as a search-results page.
# The captured New World page wraps its results section in id="search"; a
# document without it yields err(no-results-container), while a page with
# the container but zero product cards is ok with empty products.
_RESULTS_CONTAINER_ID = "search"
_NO_RESULTS_CONTAINER_CODE = "no-results-container"
_NO_RESULTS_CONTAINER_MESSAGE = 'no search results container (id="search") in document'

# HTML void elements never take a closing tag and must not be pushed onto
# the open-element stack (mirrors the html5ever behavior the Rust core gets
# for free).
_VOID_ELEMENTS = frozenset(
    {
        "area",
        "base",
        "br",
        "col",
        "embed",
        "hr",
        "img",
        "input",
        "link",
        "meta",
        "param",
        "source",
        "track",
        "wbr",
    }
)


class _Element:
    """One element node in the tolerant tree; children mix ``_Element`` and
    ``str`` (text) nodes."""

    __slots__ = ("tag", "attrs", "children")

    def __init__(self, tag: str, attrs: dict[str, str]):
        self.tag = tag
        self.attrs = attrs
        self.children: list[_Element | str] = []


class _TreeBuilder(HTMLParser):
    """Tolerant tree builder over the stdlib tokenizer: void elements are
    never pushed, a close tag pops to its nearest matching open tag, and an
    unmatched close tag is ignored."""

    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.root = _Element("#root", {})
        self._stack = [self.root]

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        # First occurrence wins for duplicated attribute names (html5ever
        # behavior); a valueless attribute normalizes to "".
        attr_dict: dict[str, str] = {}
        for name, value in attrs:
            if name not in attr_dict:
                attr_dict[name] = value if value is not None else ""
        element = _Element(tag, attr_dict)
        self._stack[-1].children.append(element)
        if tag not in _VOID_ELEMENTS:
            self._stack.append(element)

    def handle_endtag(self, tag: str) -> None:
        for index in range(len(self._stack) - 1, 0, -1):
            if self._stack[index].tag == tag:
                del self._stack[index:]
                return

    def handle_data(self, data: str) -> None:
        self._stack[-1].children.append(data)


def _parse_html(html: str) -> _Element:
    builder = _TreeBuilder()
    builder.feed(html)
    builder.close()
    return builder.root


def _iter_elements(element: _Element):
    """Every descendant element of ``element``, in document (DFS) order."""
    for child in element.children:
        if isinstance(child, _Element):
            yield child
            yield from _iter_elements(child)


def _text(element: _Element) -> str:
    """Concatenated descendant text, trimmed."""
    parts: list[str] = []

    def walk(node: _Element) -> None:
        for child in node.children:
            if isinstance(child, _Element):
                walk(child)
            else:
                parts.append(child)

    walk(element)
    return "".join(parts).strip()


def _first_by_testid(scope: _Element, testid: str) -> _Element | None:
    for element in _iter_elements(scope):
        if element.attrs.get("data-testid") == testid:
            return element
    return None


def _is_digits(text: str) -> bool:
    return bool(text) and all(char in "0123456789" for char in text)


def _find_first(text: str, chars: str) -> int:
    """Index of the first occurrence of any of ``chars``, else ``len(text)``."""
    for index, char in enumerate(text):
        if char in chars:
            return index
    return len(text)


def remove_dot_segments(path: str) -> str:
    """Remove ``.`` and ``..`` segments from an absolute URL path (RFC 3986
    section 5.2.4, simplified for already-merged paths)."""
    segments = path.split("/")
    out: list[str] = []
    for segment in segments:
        if segment == ".":
            continue
        if segment == "..":
            if len(out) > 1:
                out.pop()
            continue
        out.append(segment)
    if segments[-1] in (".", ".."):
        out.append("")
    return "/".join(out)


def _split_url(url: str) -> tuple[str, str, str]:
    """Split ``url`` into ``(origin, path, query)`` where ``origin`` is
    ``scheme://authority`` (or just ``scheme:`` for a URL without an
    authority), ``path`` excludes the query/fragment, and ``query`` excludes
    its leading ``?``."""
    colon = url.find(":")
    scheme_end = colon + 1 if colon != -1 else 0
    scheme, rest = url[:scheme_end], url[scheme_end:]
    authority = ""
    if rest.startswith("//"):
        stripped = rest[2:]
        end = _find_first(stripped, "/?#")
        authority = "//" + stripped[:end]
        rest = stripped[end:]
    path_end = _find_first(rest, "?#")
    path, tail = rest[:path_end], rest[path_end:]
    query = ""
    if tail.startswith("?"):
        after = tail[1:]
        query = after[: _find_first(after, "#")]
    return scheme + authority, path, query


def _is_absolute_url(href: str) -> bool:
    colon = href.find(":")
    if colon == -1:
        return False
    scheme = href[:colon]
    if not scheme:
        return False
    first, rest = scheme[0], scheme[1:]
    if not (first.isascii() and first.isalpha()):
        return False
    return all((char.isascii() and char.isalnum()) or char in "+.-" for char in rest)


def resolve_url(base: str, href: str) -> str:
    """Resolve ``href`` against the absolute ``base`` URL. A deliberately
    small RFC-3986-style resolver, implemented identically in every language
    core so canonical URLs match byte-for-byte across implementations."""
    if _is_absolute_url(href):
        return href
    origin, base_path, base_query = _split_url(base)
    if href.startswith("//"):
        colon = origin.find(":")
        scheme = origin[: colon + 1] if colon != -1 else ""
        return scheme + href
    if href == "" or href.startswith("#"):
        query = f"?{base_query}" if base_query else ""
        return f"{origin}{base_path}{query}{href}"
    if href.startswith("?"):
        return f"{origin}{base_path}{href}"
    split = _find_first(href, "?#")
    href_path, suffix = href[:split], href[split:]
    if href_path.startswith("/"):
        merged = href_path
    else:
        slash = base_path.rfind("/")
        dir_end = slash + 1 if slash != -1 else 0
        merged = "/" + base_path[:dir_end].lstrip("/") + href_path
    return f"{origin}{remove_dot_segments(merged)}{suffix}"


def canonical_card_url(base: str, href: str) -> str:
    """Canonicalize a product-card link: resolve against ``base``, drop any
    fragment, drop the volatile ``tr`` query parameter, and keep every other
    query parameter exactly as found."""
    resolved = resolve_url(base, href)
    without_fragment = resolved[: _find_first(resolved, "#")]
    question = without_fragment.find("?")
    if question == -1:
        return without_fragment
    before, query = without_fragment[:question], without_fragment[question + 1 :]
    kept = [pair for pair in query.split("&") if pair[: _find_first(pair, "=")] != "tr"]
    if not kept:
        return before
    return before + "?" + "&".join(kept)


def _amount_in_cents(dollars_text: str, cents_text: str) -> int | None:
    """``dollars * 100 + cents`` in cents, or ``None`` when the total does
    not fit the WIT ``amount-cents`` type (u32)."""
    amount = int(dollars_text) * 100 + int(cents_text)
    if amount > _U32_MAX:
        return None
    return amount


def _extract_price(card: _Element) -> dict | None:
    group = _first_by_testid(card, "price")
    if group is None:
        return None
    dollars_el = _first_by_testid(group, "price-dollars")
    cents_el = _first_by_testid(group, "price-cents")
    per_el = _first_by_testid(group, "price-per")
    if dollars_el is None or cents_el is None or per_el is None:
        return None
    dollars_text, cents_text = _text(dollars_el), _text(cents_el)
    if (
        not _is_digits(dollars_text)
        or len(cents_text) != 2
        or not _is_digits(cents_text)
    ):
        return None
    amount = _amount_in_cents(dollars_text, cents_text)
    if amount is None:
        return None
    return {
        "amount-cents": amount,
        "per": _text(per_el),
        "display": f"${dollars_text}.{cents_text}",
    }


def _extract_unit_price(card: _Element) -> dict | None:
    element = _first_by_testid(card, "non-promo-unit-price")
    if element is None:
        return None
    display = _text(element)
    # "$<dollars>.<2-digit cents>/<unit>" (e.g. "$0.81/ea"); anything else
    # -- promo layouts, missing decimals -- is treated as absent.
    if not display.startswith("$"):
        return None
    rest = display[1:]
    if "/" not in rest:
        return None
    amount_text, _, unit = rest.partition("/")
    if "." not in amount_text:
        return None
    dollars_text, _, cents_text = amount_text.partition(".")
    if (
        not _is_digits(dollars_text)
        or len(cents_text) != 2
        or not _is_digits(cents_text)
        or not unit
    ):
        return None
    amount = _amount_in_cents(dollars_text, cents_text)
    if amount is None:
        return None
    return {"amount-cents": amount, "unit": unit, "display": display}


def _extract_card(card: _Element, source_url: str) -> dict | None:
    # The product id comes from the data-testid on the Product element
    # itself; descendants carry product-* test ids too, so prefix-matching
    # anywhere else is wrong.
    testid = card.attrs.get("data-testid")
    if testid is None or not testid.startswith(_PRODUCT_TESTID_PREFIX):
        return None
    title_el = _first_by_testid(card, "product-title")
    if title_el is None:
        return None
    href = next(
        (
            element.attrs["href"]
            for element in _iter_elements(card)
            if element.tag == "a" and "href" in element.attrs
        ),
        None,
    )
    if href is None:
        return None
    subtitle_el = _first_by_testid(card, "product-subtitle")
    image_el = _first_by_testid(card, "product-image")
    image_src = image_el.attrs.get("src") if image_el is not None else None
    return {
        "product-id": testid[len(_PRODUCT_TESTID_PREFIX) :],
        "name": _text(title_el),
        "subtitle": _text(subtitle_el) if subtitle_el is not None else None,
        "url": canonical_card_url(source_url, href),
        "image-url": resolve_url(source_url, image_src)
        if image_src is not None
        else None,
        "price": _extract_price(card),
        "unit-price": _extract_unit_price(card),
    }


def _find_next_page_url(root: _Element, source_url: str) -> str | None:
    for element in _iter_elements(root):
        if element.tag != "a":
            continue
        rel = element.attrs.get("rel")
        if rel is None:
            continue
        if not any(token.lower() == "next" for token in rel.split()):
            continue
        href = element.attrs.get("href")
        if href is None:
            continue
        return resolve_url(source_url, href)
    return None


def parse_search_results(html: str, source_url: str) -> dict:
    """Parse one New World shop-search results page into structured
    product-card data, returned as the harness-wide result envelope. See
    ``common/wit/html-parser.wit`` for the contract and
    docs/html-parser-plan.md for the pinned observable behavior."""
    root = _parse_html(html)

    if not any(
        element.attrs.get("id") == _RESULTS_CONTAINER_ID
        for element in _iter_elements(root)
    ):
        return {
            "err": {
                "code": _NO_RESULTS_CONTAINER_CODE,
                "message": _NO_RESULTS_CONTAINER_MESSAGE,
            }
        }

    products: list[dict] = []
    seen_urls: set[str] = set()
    for card in _iter_elements(root):
        if card.attrs.get("itemtype") != _PRODUCT_ITEMTYPE:
            continue
        product = _extract_card(card, source_url)
        if product is None or product["url"] in seen_urls:
            continue
        seen_urls.add(product["url"])
        products.append(product)

    return {
        "ok": {
            "site": _SITE,
            "source-url": source_url,
            "next-page-url": _find_next_page_url(root, source_url),
            "products": products,
        }
    }

File: javascript/component/src/new-world-parser.js
// Parsing core for the `new-world-product-search` WIT interface
// (common/wit/html-parser.wit).
//
// One parsing core per language (docs/html-parser-plan.md): this module is
// used directly by the host-side native tests in tests/ and compiled
// unchanged into new-world-parser.wasm. The pure extraction logic and the
// thin WIT-binding glue (the `newWorldProductSearch` export at the end of
// this file) share one module because componentize-js evaluates a single
// self-contained entry and does not resolve relative imports -- there is
// never a second copy of the extraction logic in this language.
//
// The HTML tree is built with a small hand-rolled tolerant tokenizer rather
// than a third-party parser (parse5): constitution section 6.3 keeps the
// component pure and dependency-free, jco componentize only has to bundle a
// single self-contained ESM module, and porting the exact same
// tokenizer the Python core uses (stdlib html.parser) keeps the structured
// output -- and therefore the deduplication keys -- byte-identical across
// all three implementations. `parse_search_results` returns the harness-wide
// result envelope directly: `{ok: {...}}` or `{err: {...}}`, every record a
// plain object keyed exactly like the JSON contract (kebab-case, `null` for
// an absent WIT `option<>`).
//
// URL resolution is a small explicit RFC-3986-style resolver implemented
// identically in all three language cores (rather than each language's URL
// library) so canonical URLs -- and therefore dedup keys -- match
// byte-for-byte across implementations.

const SITE = "new-world";
const PRODUCT_ITEMTYPE = "https://schema.org/Product";
const PRODUCT_TESTID_PREFIX = "product-";
const U32_MAX = 4294967295;

// The element that makes a document recognizable as a search-results page.
// The captured New World page wraps its results section in id="search"; a
// document without it yields err(no-results-container), while a page with
// the container but zero product cards is ok with empty products.
const RESULTS_CONTAINER_ID = "search";
const NO_RESULTS_CONTAINER_CODE = "no-results-container";
const NO_RESULTS_CONTAINER_MESSAGE =
  'no search results container (id="search") in document';

// Void elements never take a closing tag and must not be pushed onto the
// open-element stack (mirrors the html5ever/html.parser behavior).
const VOID_ELEMENTS = new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr",
]);

// Raw-text elements: their content is text, never markup, so the tokenizer
// consumes everything up to the matching close tag without parsing tags
// inside (matches html.parser's CDATA handling; keeps the big inline Next.js
// <script> JSON from corrupting the element tree).
const RAWTEXT_ELEMENTS = new Set(["script", "style", "textarea", "title"]);

const NAMED_ENTITIES = {
  amp: "&",
  lt: "<",
  gt: ">",
  quot: '"',
  apos: "'",
  nbsp: " ",
};

function decodeEntities(text) {
  if (!text.includes("&")) {
    return text;
  }
  return text.replace(
    /&(#[xX][0-9a-fA-F]+|#[0-9]+|[a-zA-Z]+);/g,
    (match, body) => {
      if (body[0] === "#") {
        const code =
          body[1] === "x" || body[1] === "X"
            ? parseInt(body.slice(2), 16)
            : parseInt(body.slice(1), 10);
        return String.fromCodePoint(code);
      }
      const named = NAMED_ENTITIES[body.toLowerCase()];
      return named === undefined ? match : named;
    },
  );
}

function makeElement(tag, attrs) {
  return { tag, attrs, children: [] };
}

// Parse a start tag beginning at `html[start]` === "<". Returns
// {tag, attrs, selfClose, end} where `end` is the index just past ">", or
// null when this "<" does not begin a well-formed start tag (treated as
// text by the caller).
function matchStartTag(html, start) {
  const nameMatch = /^<([a-zA-Z][^\s/>]*)/.exec(html.slice(start));
  if (nameMatch === null) {
    return null;
  }
  const tag = nameMatch[1].toLowerCase();
  let i = start + nameMatch[0].length;
  const attrs = {};
  let selfClose = false;
  while (i < html.length) {
    while (i < html.length && /\s/.test(html[i])) {
      i += 1;
    }
    if (i >= html.length) {
      break;
    }
    if (html[i] === ">") {
      i += 1;
      return { tag, attrs, selfClose, end: i };
    }
    if (html[i] === "/") {
      selfClose = true;
      i += 1;
      continue;
    }
    let name = "";
    while (i < html.length && !/[\s/>=]/.test(html[i])) {
      name += html[i];
      i += 1;
    }
    while (i < html.length && /\s/.test(html[i])) {
      i += 1;
    }
    let value = "";
    if (html[i] === "=") {
      i += 1;
      while (i < html.length && /\s/.test(html[i])) {
        i += 1;
      }
      if (html[i] === '"' || html[i] === "'") {
        const quote = html[i];
        i += 1;
        const close = html.indexOf(quote, i);
        const stop = close === -1 ? html.length : close;
        value = decodeEntities(html.slice(i, stop));
        i = close === -1 ? html.length : close + 1;
      } else {
        let raw = "";
        while (i < html.length && !/[\s>]/.test(html[i])) {
          raw += html[i];
          i += 1;
        }
        value = decodeEntities(raw);
      }
    }
    // First occurrence wins for a duplicated attribute name; a valueless
    // attribute normalizes to "".
    if (name !== "" && !(name.toLowerCase() in attrs)) {
      attrs[name.toLowerCase()] = value;
    }
  }
  // Unterminated start tag: consume to end of input.
  return { tag, attrs, selfClose, end: html.length };
}

function indexOfCloseTag(html, tag, from) {
  const lower = html.toLowerCase();
  const needle = `</${tag}`;
  return lower.indexOf(needle, from);
}

export function parseHtml(html) {
  const root = makeElement("#root", {});
  const stack = [root];
  let i = 0;
  while (i < html.length) {
    if (html[i] !== "<") {
      const next = html.indexOf("<", i);
      const end = next === -1 ? html.length : next;
      stack[stack.length - 1].children.push(decodeEntities(html.slice(i, end)));
      i = end;
      continue;
    }
    if (html.startsWith("<!--", i)) {
      const end = html.indexOf("-->", i + 4);
      i = end === -1 ? html.length : end + 3;
      continue;
    }
    if (html.startsWith("<!", i) || html.startsWith("<?", i)) {
      const end = html.indexOf(">", i);
      i = end === -1 ? html.length : end + 1;
      continue;
    }
    if (html.startsWith("</", i)) {
      const end = html.indexOf(">", i);
      if (end === -1) {
        i = html.length;
        continue;
      }
      const name = html
        .slice(i + 2, end)
        .trim()
        .toLowerCase();
      for (let s = stack.length - 1; s > 0; s -= 1) {
        if (stack[s].tag === name) {
          stack.length = s;
          break;
        }
      }
      i = end + 1;
      continue;
    }
    const match = matchStartTag(html, i);
    if (match === null) {
      stack[stack.length - 1].children.push("<");
      i += 1;
      continue;
    }
    const element = makeElement(match.tag, match.attrs);
    stack[stack.length - 1].children.push(element);
    if (VOID_ELEMENTS.has(match.tag) || match.selfClose) {
      i = match.end;
      continue;
    }
    if (RAWTEXT_ELEMENTS.has(match.tag)) {
      const close = indexOfCloseTag(html, match.tag, match.end);
      const rawEnd = close === -1 ? html.length : close;
      const raw = html.slice(match.end, rawEnd);
      if (raw !== "") {
        element.children.push(raw);
      }
      if (close === -1) {
        i = html.length;
      } else {
        const closeEnd = html.indexOf(">", close);
        i = closeEnd === -1 ? html.length : closeEnd + 1;
      }
      continue;
    }
    stack.push(element);
    i = match.end;
  }
  return root;
}

function* iterElements(element) {
  for (const child of element.children) {
    if (typeof child !== "string") {
      yield child;
      yield* iterElements(child);
    }
  }
}

function textOf(element) {
  const parts = [];
  const walk = (node) => {
    for (const child of node.children) {
      if (typeof child === "string") {
        parts.push(child);
      } else {
        walk(child);
      }
    }
  };
  walk(element);
  return parts.join("").trim();
}

function firstByTestid(scope, testid) {
  for (const element of iterElements(scope)) {
    if (element.attrs["data-testid"] === testid) {
      return element;
    }
  }
  return null;
}

function isDigits(text) {
  return text.length > 0 && /^[0-9]+$/.test(text);
}

function findFirst(text, chars) {
  for (let index = 0; index < text.length; index += 1) {
    if (chars.includes(text[index])) {
      return index;
    }
  }
  return text.length;
}

export function removeDotSegments(path) {
  const segments = path.split("/");
  const out = [];
  for (const segment of segments) {
    if (segment === ".") {
      continue;
    }
    if (segment === "..") {
      if (out.length > 1) {
        out.pop();
      }
      continue;
    }
    out.push(segment);
  }
  const last = segments[segments.length - 1];
  if (last === "." || last === "..") {
    out.push("");
  }
  return out.join("/");
}

function splitUrl(url) {
  const colon = url.indexOf(":");
  const schemeEnd = colon === -1 ? 0 : colon + 1;
  const scheme = url.slice(0, schemeEnd);
  let rest = url.slice(schemeEnd);
  let authority = "";
  if (rest.startsWith("//")) {
    const stripped = rest.slice(2);
    const end = findFirst(stripped, "/?#");
    authority = "//" + stripped.slice(0, end);
    rest = stripped.slice(end);
  }
  const pathEnd = findFirst(rest, "?#");
  const path = rest.slice(0, pathEnd);
  const tail = rest.slice(pathEnd);
  let query = "";
  if (tail.startsWith("?")) {
    const after = tail.slice(1);
    query = after.slice(0, findFirst(after, "#"));
  }
  return { origin: scheme + authority, path, query };
}

function isAbsoluteUrl(href) {
  const colon = href.indexOf(":");
  if (colon <= 0) {
    return false;
  }
  const scheme = href.slice(0, colon);
  if (!/^[a-zA-Z][a-zA-Z0-9+.-]*$/.test(scheme)) {
    return false;
  }
  return true;
}

export function resolveUrl(base, href) {
  if (isAbsoluteUrl(href)) {
    return href;
  }
  const { origin, path: basePath, query: baseQuery } = splitUrl(base);
  if (href.startsWith("//")) {
    const colon = origin.indexOf(":");
    const scheme = colon === -1 ? "" : origin.slice(0, colon + 1);
    return scheme + href;
  }
  if (href === "" || href.startsWith("#")) {
    const query = baseQuery === "" ? "" : `?${baseQuery}`;
    return `${origin}${basePath}${query}${href}`;
  }
  if (href.startsWith("?")) {
    return `${origin}${basePath}${href}`;
  }
  const split = findFirst(href, "?#");
  const hrefPath = href.slice(0, split);
  const suffix = href.slice(split);
  let merged;
  if (hrefPath.startsWith("/")) {
    merged = hrefPath;
  } else {
    const slash = basePath.lastIndexOf("/");
    const dirEnd = slash === -1 ? 0 : slash + 1;
    merged = "/" + basePath.slice(0, dirEnd).replace(/^\/+/, "") + hrefPath;
  }
  return `${origin}${removeDotSegments(merged)}${suffix}`;
}

export function canonicalCardUrl(base, href) {
  const resolved = resolveUrl(base, href);
  const withoutFragment = resolved.slice(0, findFirst(resolved, "#"));
  const question = withoutFragment.indexOf("?");
  if (question === -1) {
    return withoutFragment;
  }
  const before = withoutFragment.slice(0, question);
  const query = withoutFragment.slice(question + 1);
  const kept = query
    .split("&")
    .filter((pair) => pair.slice(0, findFirst(pair, "=")) !== "tr");
  if (kept.length === 0) {
    return before;
  }
  return before + "?" + kept.join("&");
}

function amountInCents(dollarsText, centsText) {
  const amount = parseInt(dollarsText, 10) * 100 + parseInt(centsText, 10);
  if (amount > U32_MAX) {
    return null;
  }
  return amount;
}

function extractPrice(card) {
  const group = firstByTestid(card, "price");
  if (group === null) {
    return null;
  }
  const dollarsEl = firstByTestid(group, "price-dollars");
  const centsEl = firstByTestid(group, "price-cents");
  const perEl = firstByTestid(group, "price-per");
  if (dollarsEl === null || centsEl === null || perEl === null) {
    return null;
  }
  const dollarsText = textOf(dollarsEl);
  const centsText = textOf(centsEl);
  if (
    !isDigits(dollarsText) ||
    centsText.length !== 2 ||
    !isDigits(centsText)
  ) {
    return null;
  }
  const amount = amountInCents(dollarsText, centsText);
  if (amount === null) {
    return null;
  }
  return {
    "amount-cents": amount,
    per: textOf(perEl),
    display: `$${dollarsText}.${centsText}`,
  };
}

function extractUnitPrice(card) {
  const element = firstByTestid(card, "non-promo-unit-price");
  if (element === null) {
    return null;
  }
  const display = textOf(element);
  // "$<dollars>.<2-digit cents>/<unit>" (e.g. "$0.81/ea"); anything else --
  // promo layouts, missing decimals -- is treated as absent.
  if (!display.startsWith("$")) {
    return null;
  }
  const rest = display.slice(1);
  if (!rest.includes("/")) {
    return null;
  }
  const slash = rest.indexOf("/");
  const amountText = rest.slice(0, slash);
  const unit = rest.slice(slash + 1);
  if (!amountText.includes(".")) {
    return null;
  }
  const dot = amountText.indexOf(".");
  const dollarsText = amountText.slice(0, dot);
  const centsText = amountText.slice(dot + 1);
  if (
    !isDigits(dollarsText) ||
    centsText.length !== 2 ||
    !isDigits(centsText) ||
    unit === ""
  ) {
    return null;
  }
  const amount = amountInCents(dollarsText, centsText);
  if (amount === null) {
    return null;
  }
  return { "amount-cents": amount, unit, display };
}

function firstHref(card) {
  for (const element of iterElements(card)) {
    if (element.tag === "a" && "href" in element.attrs) {
      return element.attrs.href;
    }
  }
  return null;
}

function extractCard(card, sourceUrl) {
  // The product id comes from the data-testid on the Product element itself;
  // descendants carry product-* test ids too, so prefix-matching anywhere
  // else is wrong.
  const testid = card.attrs["data-testid"];
  if (testid === undefined || !testid.startsWith(PRODUCT_TESTID_PREFIX)) {
    return null;
  }
  const titleEl = firstByTestid(card, "product-title");
  if (titleEl === null) {
    return null;
  }
  const href = firstHref(card);
  if (href === null) {
    return null;
  }
  const subtitleEl = firstByTestid(card, "product-subtitle");
  const imageEl = firstByTestid(card, "product-image");
  const imageSrc = imageEl === null ? undefined : imageEl.attrs.src;
  return {
    "product-id": testid.slice(PRODUCT_TESTID_PREFIX.length),
    name: textOf(titleEl),
    subtitle: subtitleEl === null ? null : textOf(subtitleEl),
    url: canonicalCardUrl(sourceUrl, href),
    "image-url":
      imageSrc === undefined ? null : resolveUrl(sourceUrl, imageSrc),
    price: extractPrice(card),
    "unit-price": extractUnitPrice(card),
  };
}

function findNextPageUrl(root, sourceUrl) {
  for (const element of iterElements(root)) {
    if (element.tag !== "a") {
      continue;
    }
    const rel = element.attrs.rel;
    if (rel === undefined) {
      continue;
    }
    if (!rel.split(/\s+/).some((token) => token.toLowerCase() === "next")) {
      continue;
    }
    const href = element.attrs.href;
    if (href === undefined) {
      continue;
    }
    return resolveUrl(sourceUrl, href);
  }
  return null;
}

// Parse one New World shop-search results page into structured product-card
// data, returned as the harness-wide result envelope. See
// common/wit/html-parser.wit for the contract and docs/html-parser-plan.md
// for the pinned observable behavior.
export function parseSearchResults(html, sourceUrl) {
  const root = parseHtml(html);

  let hasContainer = false;
  for (const element of iterElements(root)) {
    if (element.attrs.id === RESULTS_CONTAINER_ID) {
      hasContainer = true;
      break;
    }
  }
  if (!hasContainer) {
    return {
      err: {
        code: NO_RESULTS_CONTAINER_CODE,
        message: NO_RESULTS_CONTAINER_MESSAGE,
      },
    };
  }

  const products = [];
  const seenUrls = new Set();
  for (const card of iterElements(root)) {
    if (card.attrs.itemtype !== PRODUCT_ITEMTYPE) {
      continue;
    }
    const product = extractCard(card, sourceUrl);
    if (product === null || seenUrls.has(product.url)) {
      continue;
    }
    seenUrls.add(product.url);
    products.push(product);
  }

  return {
    ok: {
      site: SITE,
      "source-url": sourceUrl,
      "next-page-url": findNextPageUrl(root, sourceUrl),
      products,
    },
  };
}

// ---------------------------------------------------------------------------
// Component binding glue for the `new-world-parser` world.
//
// This is the only JavaScript-binding-specific layer: it adapts the
// language-neutral envelope above to the shapes jco/componentize-js expects
// for the WIT export. Unlike Python (componentize-py `-p src` bundles a whole
// directory) and Rust (a path-dependency crate), componentize-js evaluates a
// single self-contained module and does not resolve relative imports -- so
// the core and this thin glue share one file (docs/html-parser-plan.md: core
// placement is decided per language by what the component toolchain can
// bundle). jco maps interface `new-world-product-search` to the named export
// `newWorldProductSearch`, its function `parse-search-results` to
// `parseSearchResults(html, sourceUrl)`, and WIT record fields to camelCase
// properties. For a `result<T, E>` return the guest returns the ok value and
// THROWS the error value (symmetric with the Python parser_app raising for
// err); `option<T>` none is represented as `undefined`.
// ---------------------------------------------------------------------------

function toPrice(price) {
  if (price === null) {
    return undefined;
  }
  return {
    amountCents: price["amount-cents"],
    per: price.per,
    display: price.display,
  };
}

function toUnitPrice(unitPrice) {
  if (unitPrice === null) {
    return undefined;
  }
  return {
    amountCents: unitPrice["amount-cents"],
    unit: unitPrice.unit,
    display: unitPrice.display,
  };
}

function toProductCard(product) {
  return {
    productId: product["product-id"],
    name: product.name,
    subtitle: product.subtitle === null ? undefined : product.subtitle,
    url: product.url,
    imageUrl: product["image-url"] === null ? undefined : product["image-url"],
    price: toPrice(product.price),
    unitPrice: toUnitPrice(product["unit-price"]),
  };
}

function toSearchResults(ok) {
  return {
    site: ok.site,
    sourceUrl: ok["source-url"],
    nextPageUrl: ok["next-page-url"] === null ? undefined : ok["next-page-url"],
    products: ok.products.map(toProductCard),
  };
}

export const newWorldProductSearch = {
  parseSearchResults(html, sourceUrl) {
    const envelope = parseSearchResults(html, sourceUrl);
    if (envelope.err !== undefined) {
      throw { code: envelope.err.code, message: envelope.err.message };
    }
    return toSearchResults(envelope.ok);
  },
};
